### ISSUE TITLE
rad2deg/deg2rad: float32 kornia.constants.pi is cast to the input dtype — integer inputs divide by 3, float64 loses ~7 digits

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#3937]
`kornia.geometry.conversions.rad2deg` and `deg2rad` divide by `kornia.constants.pi` after casting that constant to the **input** tensor's dtype (`conversions.py:97` and `:127`: `180.0 * tensor / pi.to(tensor.device).type(tensor.dtype)`). `pi` is created as `torch.tensor(3.14159265358979323846)` (`kornia/constants.py:25`), which materialises as a **float32** tensor. Two consequences follow: for an **integer** input tensor the constant is cast to int64 and becomes literally `3`, so the conversion is off by ~5 %; for a **float64** input the constant is upcast from its already-truncated float32 value, so ~7 significant digits are silently lost. `angle_to_rotation_matrix` calls `deg2rad` and inherits the integer bug directly. `torch.rad2deg` / `torch.deg2rad` are correct in both cases.

## Repro

```python
import math
import torch
from kornia.constants import pi
from kornia.geometry.conversions import rad2deg, deg2rad, angle_to_rotation_matrix

print("pi.dtype, pi.item():", pi.dtype, pi.item())
print("pi.type(torch.int64):", pi.type(torch.int64))
print("pi.type(torch.float64).item():", pi.type(torch.float64).item(), "| math.pi:", math.pi)

# (1) integer input
print("kornia rad2deg int:", rad2deg(torch.tensor([1, 2, 3])))
print("torch  rad2deg int:", torch.rad2deg(torch.tensor([1, 2, 3])))
print("kornia deg2rad int:", deg2rad(torch.tensor([180, 90])))
print("torch  deg2rad int:", torch.deg2rad(torch.tensor([180, 90])))

# (2) downstream: angle_to_rotation_matrix with an integer angle
print("angle_to_rotation_matrix(tensor([90])) :", angle_to_rotation_matrix(torch.tensor([90])))
print("angle_to_rotation_matrix(tensor([90.])):", angle_to_rotation_matrix(torch.tensor([90.])))

# (3) float64 precision
x64 = torch.tensor(math.pi, dtype=torch.float64)
print("kornia rad2deg f64 err:", rad2deg(x64).item() - 180.0)
print("torch  rad2deg f64 err:", torch.rad2deg(x64).item() - 180.0)
d64 = torch.tensor(180.0, dtype=torch.float64)
print("kornia deg2rad f64:", deg2rad(d64).item(), "| torch:", torch.deg2rad(d64).item(), "| math.pi:", math.pi)
print("kornia deg2rad f64 abs err:", abs(deg2rad(d64).item() - math.pi))
```

## Observed

```
pi.dtype, pi.item(): torch.float32 3.1415927410125732
pi.type(torch.int64): tensor(3)
pi.type(torch.float64).item(): 3.1415927410125732 | math.pi: 3.141592653589793
kornia rad2deg int: tensor([ 60., 120., 180.])
torch  rad2deg int: tensor([ 57.2958, 114.5916, 171.8873])
kornia deg2rad int: tensor([3.0000, 1.5000])
torch  deg2rad int: tensor([3.1416, 1.5708])
angle_to_rotation_matrix(tensor([90])) : tensor([[[ 0.0707,  0.9975],
         [-0.9975,  0.0707]]])
angle_to_rotation_matrix(tensor([90.])): tensor([[[-4.3711e-08,  1.0000e+00],
         [-1.0000e+00, -4.3711e-08]]])
kornia rad2deg f64 err: -5.008956179608504e-06
torch  rad2deg f64 err: 0.0
kornia deg2rad f64: 3.1415927410125732 | torch: 3.141592653589793 | math.pi: 3.141592653589793
kornia deg2rad f64 abs err: 8.742278012618954e-08
```

## Expected

- `rad2deg(torch.tensor([1, 2, 3]))` should be `[57.2958, 114.5916, 171.8873]`, matching `torch.rad2deg`. Observed `[60., 120., 180.]` is exactly what you get from dividing by 3 instead of π — the output dtype is already promoted to float32, so nothing about the integer input forces an integer constant.
- `angle_to_rotation_matrix(torch.tensor([90]))` should equal `angle_to_rotation_matrix(torch.tensor([90.]))` (the quarter-turn matrix). Observed is the matrix for `90 * 3 / 180 = 1.5` rad ≈ 85.94°.
- `deg2rad(torch.tensor(180., dtype=torch.float64))` should be `3.141592653589793` (what `torch.deg2rad` returns). Observed `3.1415927410125732` carries an absolute error of `8.74e-08` — about **7 significant digits of the 16 that float64 provides**. Symmetrically `rad2deg(π, f64)` is off 180° by `-5.0e-06` where `torch.rad2deg` is exact.

Note that `rad2deg(deg2rad(x))` still round-trips cleanly, because the same biased constant appears in the numerator and denominator and cancels — which is why this has gone unnoticed.

Suggested fix: delegate to `torch.rad2deg` / `torch.deg2rad`, or at minimum build the constant at the input's dtype from a float64 source rather than casting a float32 tensor.

To size the blast radius on the common float32 path, here is the exact comparison snippet (fixed seed, four input ranges spanning six orders of magnitude — absolute deltas are scale-dependent and not meaningful here, so this reports ULP distance and relative error):

```python
import torch
from kornia.geometry.conversions import rad2deg, deg2rad

def cmp(name, fn_k, fn_t, x):
    a, b = fn_k(x), fn_t(x)
    same = (a == b).float().mean().item()
    rel = (a - b).abs() / b.abs().clamp(min=torch.finfo(torch.float32).tiny)
    ulp = (a.view(torch.int32) - b.view(torch.int32)).abs().max().item()
    print(f"{name:>24}: bit-identical {same * 100:5.1f}%  max rel {rel.max().item():.3e}  max ULP {ulp}")

g = torch.Generator().manual_seed(0)
ranges = {
    "U(-10,10)": torch.rand(100000, generator=g) * 20 - 10,
    "U(-1e3,1e3)": torch.rand(100000, generator=g) * 2e3 - 1e3,
    "U(-1e6,1e6)": torch.rand(100000, generator=g) * 2e6 - 1e6,
    "U(0,1)": torch.rand(100000, generator=g),
}
for n, x in ranges.items():
    cmp("rad2deg " + n, rad2deg, torch.rad2deg, x)
for n, x in ranges.items():
    cmp("deg2rad " + n, deg2rad, torch.deg2rad, x)
```

```
       rad2deg U(-10,10): bit-identical  52.2%  max rel 1.869e-07  max ULP 2
     rad2deg U(-1e3,1e3): bit-identical  52.1%  max rel 1.869e-07  max ULP 2
     rad2deg U(-1e6,1e6): bit-identical  52.2%  max rel 1.867e-07  max ULP 2
          rad2deg U(0,1): bit-identical  51.9%  max rel 1.864e-07  max ULP 2
       deg2rad U(-10,10): bit-identical  57.1%  max rel 1.675e-07  max ULP 2
     deg2rad U(-1e3,1e3): bit-identical  55.6%  max rel 1.673e-07  max ULP 2
     deg2rad U(-1e6,1e6): bit-identical  55.2%  max rel 1.676e-07  max ULP 2
          deg2rad U(0,1): bit-identical  55.9%  max rel 1.676e-07  max ULP 2
```

So on float32 roughly half the values stay bit-identical and the rest move by **at most 2 ULP** (max relative error ~1.9e-07, i.e. about one float32 epsilon) — stable across all four ranges. Meanwhile the integer and float64 paths go from wrong to correct.

## Classification

**Clearly-broken-output bug**, not a semantic/default-convention change. The integer path returns numbers that are wrong by ~5 % under any reading of "convert radians to degrees" — there is no convention under which π equals 3 — and the float64 path is a plain precision defect against an exact reference. Per the reference-core lifecycle (Part I), this class may land early as an ordinary bugfix rather than waiting for the coordinated Repair & Deprecation window. The ≤2 ULP float32 movement measured above is the only behavior change on the common path.

## Environment

kornia 0.9.0rc1 @ `3358ef16`, torch 2.9.1, cpu.

Posted on behalf of @ducha-aiki by Claude (Fable 5).

**Comment by ducha-aiki** [comment]
**Current consumer inventory of `kornia.constants.pi`** — PR #3942's `rad2deg` warning now points here instead of enumerating consumers inline (an inline list already went stale once inside a single review cycle). Grep-verified at `main`-equivalent head today:

```
$ grep -rn "from kornia.constants import" kornia/ --include="*.py" | grep "pi"
kornia/__init__.py:54
kornia/augmentation/_2d/intensity/hue.py:25
kornia/augmentation/_2d/intensity/color_jitter.py:25
kornia/augmentation/_2d/intensity/color_jiggle.py:24
kornia/enhance/jpeg.py:28
kornia/feature/orientation.py:24
kornia/feature/mkd.py:24
kornia/feature/integrated.py:25
kornia/geometry/conversions.py:25
```

Two distinct defects for whoever picks this up (established during #3942's review, all claims executed):

1. **float64 precision loss** — originates in the float32 constant itself and affects every consumer above. `rad2deg(torch.tensor(math.pi, dtype=torch.float64))` is `180.0 - 5.0e-06`: the result is left with only about seven correct significant digits.
2. **Integer truncation** — originates in `rad2deg`/`deg2rad`'s own cast of the constant to the *input* dtype, not in the constant: even a float64 `constants.pi` still gives `pi.type(torch.int64) == 3`, so `rad2deg(torch.tensor([1, 2, 3])) == [60., 120., 180.]` and `deg2rad(torch.tensor([180, 90])) == [3.0, 1.5]`. A fix that only replaces the constant does not touch this; a fix that only patches `rad2deg`/`deg2rad` leaves defect 1 in the other eight consumers.

A consequence of defect 2 worth knowing: `rad2deg(deg2rad(torch.tensor([180, 90])))` returns `[171.8873, 85.9437]` — the two casts use different π factors (int-truncated 3 on the way in, float32 π on the way out), so the pair is not an inverse for integer input.

Regression pins already in place on #3942's branch: a float64-hardcoded strict xfail (`raises=AssertionError`, XPASSes loudly when defect 1 is fixed for the conversions) and a wart pin of the integer `[60., 120., 180.]` outputs (flips loudly under any handling change for defect 2).

Posted on behalf of @ducha-aiki by Claude (Fable 5).
